### PR TITLE
Unparse map to sql

### PR DESCRIPTION
**Pull request by delamarch3**
## Which issue does this PR close?

Implements `map` for #13478. For `get_field` it doesn't look like there is a way to distinguish if it is a map access from the expression.

## Rationale for this change

Support roundtrip for statements containing `map`

## What changes are included in this PR?
Implement `expr_to_sql` for `map`. 

## Are these changes tested?
Yes

## Are there any user-facing changes?
No

### CODE REVIEW

File: datafusion/functions-nested/src/map.rs [hunk 1]
```diff
@@ -214,9 +214,9 @@ impl ScalarUDFImpl for MapFunc {
     }
 
     fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
-        if arg_types.len() % 2 != 0 {
+        if arg_types.len() != 2 {
```
Comment (alamb) [inline datafusion/functions-nested/src/map.rs:217]: 👍

File: datafusion/sql/src/unparser/expr.rs [hunk 2]
```diff
@@ -567,6 +568,43 @@ impl Unparser<'_> {
...
+
+        let keys = match self.expr_to_sql(&args[0])? {
+            ast::Expr::Array(Array { elem, .. }) => elem,
+            other => {
+                return internal_err!(
+                    "map expects first argument to be an array, but received: {:?}",
+                    other
+                )
+            }
+        };
```
Comment (alamb) [inline datafusion/sql/src/unparser/expr.rs:584]: You can also write this in the following way which may be more concise:

```suggestion
        let  ast::Expr::Array(Array { elem: keys, .. })) = self.expr_to_sql(&args[0])? else {
                return internal_err!(
                    "map expects first argument to be an array, but received: {:?}",
                    other
                )
        };
```

**Review by alamb — APPROVED** [review]
Thank you @delamarch3  -- this looks good to me

**Comment by delamarch3** [comment]
Thanks for the review @alamb

**Comment by alamb** [comment]
Thanks @delamarch3 !

**Pull request merged**